engine/src/state_chain_observer/sc_observer/mod.rs

```diff
@@ -525,6 +516,39 @@ where
     }.instrument(info_span!("SCObserver")).boxed()).await
 }
 
+/// Get the CFE events from the SC at the given block with version checking
+pub async fn get_cfe_events(
+	state_chain_client: Arc<impl StorageApi>,
+	block_info: BlockInfo,
+) -> Result<Vec<CfeEvent>, anyhow::Error> {
+	// First we check the version number of the CFE events
+	let cfe_events_version = state_chain_client
+		.storage_value::<pallet_cf_cfe_interface::CfeEventsVersion<Runtime>>(block_info.hash)
+		.await
+		.expect("Failed to get CFE events version");
+
+	// For now we only have one version of CFE events, so we panic if the version number is not a
+	// match.
+	if cfe_events_version != CFE_EVENTS_VERSION {
+		panic!("Cannot process CFE events at block {}. Incompatible CFE events version: {cfe_events_version}. Using version: {CFE_EVENTS_VERSION}", block_info.number);
+	}
+
+	// Getting the events without decoding yet, so in the future we can decode them with the correct
+	// struct for the detected version.
+	match state_chain_client
+		.storage_value_raw::<pallet_cf_cfe_interface::CfeEvents<Runtime>>(block_info.hash)
+		.await
+	{
+		Ok(Some(events_data)) => <Vec<CfeEvent>>::decode(&mut &events_data[..])
+			.map_err(anyhow::Error::msg)
+			.context(format!("Failed to decode events at block {}", block_info.number)),
+		Ok(None) => Ok(vec![]),
+		Err(error) => {
+			bail!("Failed to read events at block {}. {error}", block_info.number);
```

Comment:
nit: when not returning early, we use Err(anyhow::anyhow!())
```suggestion
			bail!("Failed to read events at block {}. {error}", block_info.number);
```